### ISSUE TITLE
"Bad type on operand stack" occurs when a caught exception is passed to another context

### ISSUE THREAD

**Issue by Ktlo** [Kotlin/kotlinx.coroutines#2068]
Code below raises this error at runtime.

```kotlin
fun task() {}

fun useException(e: Exception) {}

suspend fun example() {
    try {
        task()
    } catch (e: Exception) {
        withContext(Dispatchers.IO) {
            useException(e)
        }
    }
}

fun main() = runBlocking {
    example()
}
```

I tried to compile and run this snippet with version 1.3.6 and everything was fine, but on version 1.3.7 of `kotlinx-coroutines` the following exception is thrown.

```
java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    ExampleKt.example(Lkotlin/coroutines/Continuation;)Ljava/lang/Object; @104: invokespecial
  Reason:
    Type 'java/lang/Exception' (current frame, stack[3]) is not assignable to 'kotlin/jvm/internal/Ref$ObjectRef'
  Current Frame:
    bci: @104
    flags: { }
    locals: { 'kotlin/coroutines/Continuation', 'java/lang/Exception', 'java/lang/Object', 'ExampleKt$example$1', 'java/lang/Object' }
    stack: { 'kotlin/coroutines/CoroutineContext', uninitialized 98, uninitialized 98, 'java/lang/Exception', null }
  Bytecode:
    0x0000000: 2ac1 0028 9900 202a c000 284e 2db4 002c
    0x0000010: 122d 7e99 0011 2d59 b400 2c12 2d64 b500
    0x0000020: 2ca7 000c bb00 2859 2ab7 0031 4e2d b400
    0x0000030: 354d b800 3b3a 042d b400 2caa 0000 005c
    0x0000040: 0000 0000 0000 0001 0000 0015 0000 004a
    0x0000050: 2cb8 0041 00b8 0043 a700 3b4c b800 49c0
    0x0000060: 004b bb00 4d59 2b01 b700 50c0 0052 2d2d
    0x0000070: 2bb5 0055 2d04 b500 2cb8 001b 5919 04a6
    0x0000080: 0013 1904 b02d b400 55c0 0026 4c2c b800
    0x0000090: 412c 57b2 005b b0bb 005d 5912 5fb7 0062
    0x00000a0: bf                                     
  Exception Handler Table:
    bci [84, 88] => handler: 91
  Stackmap Table:
    same_frame(@36)
    append_frame(@45,Top,Top,Object[#40])
    full_frame(@80,{Object[#102],Top,Object[#4],Object[#40],Object[#4]},{})
    same_locals_1_stack_item_frame(@91,Object[#38])
    same_frame(@133)
    full_frame(@146,{Object[#102],Object[#38],Object[#4],Object[#40],Object[#4]},{Object[#4]})
    full_frame(@147,{Object[#102],Top,Object[#4],Object[#40],Object[#4]},{})
    same_frame(@151)

	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.privateGetMethodRecursive(Class.java:3048)
	at java.lang.Class.getMethod0(Class.java:3018)
	at java.lang.Class.getMethod(Class.java:1784)
	at sun.launcher.LauncherHelper.validateMainClass(LauncherHelper.java:544)
	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:526)
Error: A JNI error has occurred, please check your installation and try again
Exception in thread "main" 
```

I also tried to create other function that simply calls `withContext` and this worked fine.

```kotlin
/* ... */
suspend fun <R> suspended(context: CoroutineContext, block: suspend CoroutineScope.() -> R): R {
    return withContext(context, block)
}

suspend fun example() {
    try {
        task()
    } catch (e: Exception) {
        suspended(Dispatchers.IO) {
            useException(e)
        }
    }
}
/* ... */
```

**Environment Details**
- jdk version: 1.8.0-u251
- kotlin version: 1.3.72
- kotlinx-coroutines version: 1.3.7

**Comment by qwwdfsad** [comment]
Thanks for the self-contained reproducer!
This is https://youtrack.jetbrains.com/issue/KT-39113 and it will be fixed only in Kotlin 1.4.

As a temporary workaround, it's enough to extract `withContext` from `catch` block to a separate function

**Comment by qwwdfsad** [comment]
Fixed in Kotlin 1.4
